File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.py
import sys
from collections import deque

T = int(input())

def bfs(x, y):
    dq = deque()
    dq.append((x, y))
    visited = [0 for _ in range(n)]

    while dq:
        x, y = dq.popleft()

        # check last
        if abs(x - end_x) + abs(y - end_y) <= macju:
            print("happy")
            return
        
        for i in range(n):
            if visited[i] == 0:
                if abs(x - peon_pos[i][0]) + abs(y - peon_pos[i][1]) <= macju:
                    dq.append((peon_pos[i][0], peon_pos[i][1]))
                    visited[i] = 1
    print("sad")

for t in range(T):
    n = int(input())

    macju = 20 * 50

    start_x, start_y = map(int, input().split())
    peon_pos = []

    for i in range(n):
        x, y = map(int, input().split())
        peon_pos.append((x, y))

    end_x, end_y = map(int, input().split())
    bfs(start_x, start_y)

File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.cc
#include <stdio.h>
#include <iostream>
#include <cstdio>
#include <cstring>

using namespace std; 

// bfs - deque 구현방법
// import queue / 선언 방법 / 추가방법
#include <queue>

// struct
struct point
{
    int x; 
    int y; 
}; 

// global 
bool visited[100] = {false, }; 
point pHome;
point pEnd;
point peon[100]; 
int macju = 20 * 50; 

// 절댓값 계산 함수 구현하기
int abs(int val) {
    if (val < 0)
        return -val ; 
    return val;  
}

void dfs(int N)
{
    queue<pair<int, int>> dq; 
    dq.push({pHome.x, pHome.y}); 

    while (!dq.empty()) {
        int x = dq.front().first; 
        int y = dq.front().second; 
        dq.pop(); 

        if (abs(x - pEnd.x) + abs(y - pEnd.y) <= macju){
            cout << "happy" << endl; 
            return; 
        }

        for(int i = 0; i < N; i ++) {
            if (visited[i] == false) {
                if (abs(x - peon[i].x) + abs(y - peon[i].y) <= macju){
                    visited[i] = true; 
                    dq.push({peon[i].x, peon[i].y}); 
                }
            }
        }     
    }
    cout << "sad" << endl; 
}

int main() {

    // init
    // ios_base::sync_with_studio(0); 
    cout.tie(0), cin.tie(0); 

    int T, n; 
    cin >> T; 

    for(int i = 0; i < T; i ++) {
        cin >> n; 

        cin >> pHome.x >> pHome.y; 

        for(int i = 0; i < n; i ++) {
            cin >> peon[i].x >> peon[i].y; 
        }

        cin >> pEnd.x >> pEnd.y; 

        dfs(n);

        // bool visited[100] = {false, }; 
        // fill(visited, visited.size(), false);
        // 초기화
        memset(visited, 0, sizeof(visited));
    }

}